lib/active_merchant/billing/gateways/moneris.rb

```diff
@@ -73,15 +74,16 @@ def purchase(money, creditcard_or_datakey, options = {})
         post[:order_id] = options[:order_id]
         post[:address] = options[:billing_address] || options[:address]
         post[:crypt_type] = options[:crypt_type] || @options[:crypt_type]
+        add_external_mpi_fields(post, options)
         add_stored_credential(post, options)
-        action = if post[:cavv]
+        action = if post[:cavv] || options[:three_d_secure]
                    'cavv_purchase'
                  elsif post[:data_key].blank?
                    'purchase'
                  else
                    'res_purchase_cc'
                  end
-        commit(action, post)
+        commit(action, post, options[:three_d_secure].present?)
```

Comment:
I'd have a mild preference to pass the entire options hash into the commit method, and have the response dig into the 3ds hash, just to have all options available by default for later features if necessary. What do you think?